feat: import samples from CSV file

leonardmq opened the pull request:
## What does this PR do?

This PR adds a `Upload dataset` feature to let users import samples / runs from a CSV file, and paves the way for more formats.

Changes:
- add button on the `Datasets` page that triggers a modal
- add `POST /api/projects/{project_id}/tasks/{task_id}/runs/bulk_upload` API route
- add dataset import handling
- add tests for the core import

Here's a screen recording showing the feature - note I added some intentional sleep on the server side to demonstrate the loading state, but otherwise it completes instantaneously:


https://github.com/user-attachments/assets/db1d9d83-02af-477a-bafa-c7c31b7a44ff


### Structure

The CSV structure must have a header row at the top, and fields must be escaped where necessary; the supported fields are:
- `input`: the run input - can be serialized JSON
- `output`: the run output - can be serialized JSON
- `reasoning` (optional): if using intermediate output
- `tags` (optional): comma-separated list of tags

Example - note that the ordering of columns does not matter:
```csv
input,output,reasoning,tags
"My name is John","{""joke"": ""They call me John Wick... but my bank calls me 'insufficient funds'.""}","The user introduced themselves, so we play along with a joke on the name 'John'.","funny"
"My name is Jane","{""joke"": ""Jane? Like Tarzan's Jane? Hope you're ready for some wild adventures!""}","Making a playful reference to the famous Tarzan and Jane pairing.","hilarious"
"My name is Jack","{""joke"": ""Jack? Hope you don't lose your spot on the door like in Titanic.""}","Classic reference to 'Jack' from Titanic, a common cultural joke.","funny,movie"
"My name is Jill","{""joke"": ""Jill? Where's Jack? I hope you two didn't tumble down a hill!""}","Classic nursery rhyme reference to Jack and Jill.",
"My name is Jack","{""joke"": ""Jack? Like Jack Sparrow? Where’s the rum?""}","Pirates of the Caribbean joke for name recognition.",
```

### Tagging

In addition to the tags that may optionally be provided in the CSV file, the import will assign two tags to the `TaskRuns` created: `imported`, and `imported_{timestamp}`.

### Invalid upload

If the file provided is invalid, the dataset importer will not create any tasks, and will raise a `KilnInvalidImportFormat` exception that refers to the particular row that caused the exception to be raised. 

The `KilnInvalidImportFormat` is handled at the API controller layer, and returned to the frontend with a `422` and then frontend displays it:
<img width="646" alt="image" src="https://github.com/user-attachments/assets/a70bf382-390b-48d0-b77a-e386809f9c85" />

## Related Issues

None.

## Contributor License Agreement

I, @leonardmq, confirm that I have read and agree to the [Contributors License Agreement](https://github.com/Kiln-AI/Kiln/blob/main/CLA.md).

## Checklists

- [x] Tests have been run locally and passed
- [x] New tests have been added to any work in /lib

leonardmq commented:
I was not sure if these files belong in the `adapters/` directory.

As the `dataset_import.py` file is already a bit bloated with CSV utils and validation helpers, I was thinking about restructuring this:
- add a folder `data_import/` and have one file per importer (e.g. `csv_importer.py`, etc.)
- add an `AbstractImporter` with a `validate()` and `create_runs()` method, and have each importer be an implementation of that class (instead of each importer being just a function), which could help avoid having the fairly large `Importer` functions
- add a `data_import_registry.py` wrapper where we register the mapping from import type to the corresponding importer

In terms of the UI, I kept the logic fairly self-contained in the `upload_dataset_dialog.svelte` component so it should be easy to move around if needed.

The design idea with this is that more importers could be easily added later - both UI and backend should support that fairly easily. In terms of potential formats that could be supported, maybe the OpenAI JSONL format used for fine-tuning could be a useful one.

Let me know what you think

scosman commented:
Amazing. I'll take a look next week!

scosman commented:
UI is nice and clean and follows Kiln style. Some small UI tweaks would be good.

- The “Upload Dataset” button should be the the AppPage buttons section - it has a standard function for adding page actions with consistent UI (it will be higher up, aligned to title). 
- Let’s remove the radio button since we only support CSV
- “Selected: filename” seems redundant since you can see it right above 
- Let’s put a little helper text in the modal/dialog explaining the fields
  - “Upload a CSV to add each row to your dataset. The CSV must have a header row. The the following columns are supported: input, output, reasoning (optional), tags (optional, comma separated string).”

Looking at code next!

scosman commented on app/web_ui/src/lib/ui/upload_dataset_dialog.svelte:
```diff
@@ -0,0 +1,139 @@
+<script lang="ts">
```
Nit: we can move this into `/dataset/[project_id]/[task_id]`. The Lib folder is mostly for components we want to reuse, and this is fairly single purpose.

leonardmq replied:
✅

leonardmq commented on app/web_ui/src/lib/ui/upload_dataset_dialog.svelte:
```diff
@@ -0,0 +1,139 @@
+  let error: KilnError | null = null
+  let loading = false
+  let selected_file: File | null = null
+  let dataset_type: "csv" | null = "csv"
```
I was thinking the dropdown could be used to add support for the JSONL format that fine-tuning are often exported to (for OpenAI for example). I can remove it from the UI but keep the extensible logic on the backend in case we decide to support more formats later.

What do you think? Remove from UI but keep extensible on the backend side?

leonardmq replied:
✅ - removed from the UI and the API, but kept most of the logic on the backend

scosman commented:
FYI - I have a review in progress with a few more smaller comments. Away from laptop but will try to finish asap

scosman commented on libs/core/kiln_ai/adapters/dataset_import.py:
```diff
@@ -0,0 +1,230 @@
+
+    input: str = Field(description="The input to the model")
+    output: str = Field(description="The output of the model")
+    reasoning: str | None = Field(
```
P2: we could also add chain_of_thought.

We store these under different keys in a TaskRun intermediate_output, since they are different things.

leonardmq replied:
✅ - added `chain_of_thought` (also in the modal UI that lists out the fields)

scosman commented on libs/core/kiln_ai/adapters/dataset_import.py:
```diff
@@ -0,0 +1,230 @@
+        if not value:
+            return []
+        # Handle string values (comma-separated tags)
+        if isinstance(value, str):
```
I'd move this out of a model validator and into where this is set. A List[str] shouldn't ever be assigned a str, and would be good for type catcher to catch/fix that. Make the caller call strip/split and assign the right type.

leonardmq replied:
To be fair, the typing did catch this initially and I kept trying different things to get around it thinking there was a legitimate way to do a transform like this (😵‍💫). Unwrapping at the caller makes a lot more sense

scosman commented on libs/core/kiln_ai/adapters/dataset_import.py:
```diff
@@ -0,0 +1,230 @@
+    def __init__(self, message: str, row_number: int | None = None):
+        self.row_number = row_number
+        if row_number is not None:
+            message = f"Error in row {row_number}: {message}"
```
👍

scosman commented on libs/core/kiln_ai/adapters/dataset_import.py:
```diff
@@ -0,0 +1,230 @@
+    try:
+        validated_row = CSVRowSchema.model_validate(row)
+    except ValidationError as e:
+        logger.warning(f"Invalid row {row_number}: {row}", exc_info=True)
```
we have this same `except ValidationError` logic twice. Remove it from this method, wrap the call to create_task_run_from_csv_row with a try, and put this exception logic there? If we add future exceptions (or accidental) in this method, it would catch them as well. No need for any try/except in this method.

This also means we don't need row_number as a parameter (only the caller needs to know about the context).

leonardmq replied:
✅ - makes a lot more sense; I think the original intent was to figure out if the error was from the CSV format itself that was problematic, or the actual semantics of it (rejected by the task run creation validation), but I guess we get it in the stack trace anyway

scosman commented on libs/core/kiln_ai/adapters/dataset_import.py:
```diff
@@ -0,0 +1,230 @@
+            parent=task,
+            input=validated_row.input,
+            input_source=DataSource(
+                type=DataSourceType.file_import,
```
We have validators for the DataSource mode that validate properties. We should add file_name there and make it required, and also make the other properties not_allowed_for DataSourceType.file_import

leonardmq replied:
✅

scosman commented on libs/server/kiln_server/run_api.py:
```diff
@@ -257,6 +274,61 @@ async def edit_tags(
+                ),
+            )
+            imported_count = importer.create_runs_from_file()
+        except KilnInvalidImportFormat as e:
```
Can we just remove the try/except here?

We can let the other exception bubble up. FastAPI will catch exceptions, log and return a 500. Nice to keep the code short/clean and get same result.

I think it will also catch the 422 automatically before reaching this since we use an enum field (it validates input and raises an error). But either way the exception would bubble up and be logged/returned without the extra code.

leonardmq replied:
The intent here was to translate the `KilnInvalidImportFormat` error into a 422. This error is raised during `import_csv(...)` if the file structure or a row fails parsing - so it is different from the `DatasetImportFormat` enum validation in the route (assuming that is the enum you were referring to).

But I should remove the generic `except Exception as e` further down, since FastAPI will handle it as a 500, like you mentioned. Were you suggesting just that, or removing the entire try/catch block? If we remove the whole try/catch entirely, invalid file parsing errors would bubble up as a 500, which might not be ideal.

scosman replied:
Keeping the 422 makes sense. Returning a 4xx when the inputs are wrong is a good call.

leonardmq replied:
✅

leonardmq commented on libs/server/kiln_server/run_api.py:
```diff
@@ -257,6 +274,61 @@ async def edit_tags(
+    async def bulk_upload(
+        project_id: str,
+        task_id: str,
+        dataset_type: DatasetImportFormat = Form(...),
```
I have tried a few combinations but could not get the client SDK to work. OpenAPI describes the file as `string` with `binary` format. The `openapi-typescript` codegen expects a `string` for such schemas. 

According to [this issue](https://github.com/openapi-ts/openapi-typescript/issues/1214), using their `transform` API to specify how to type-map a multipart form upload seems to be the recommended approach. The docs about this here: https://openapi-ts.dev/node#example-blob-types . It seems like it would require some changes to how the codegen is currently being used as we use it as CLI in `generate_schema.sh` and the dynamic `transform` looks like it runs differently from the CLI.

If we want the API to be truly JSON, we could also serialize the file to base64, but it seems inefficient as files may be large

scosman replied:
Makes sense. No need to spend any more time on this. This stuff is always a bit of a pain. Thanks for trying

leonardmq commented on libs/core/kiln_ai/adapters/dataset_import.py:
```diff
@@ -0,0 +1,230 @@
+import csv
```
Just an FYI - moving this to `utils/` creates a circular import when exporting it in `utils/__init__.py`, since `basemodel.py` depends on `utils`. So, I still moved `dataset_import.py` into `utils/`, but left it out of the `__init__.py` to avoid the issue. Let me know if this works for you, or if you’d prefer we put it elsewhere. It might be best to keep `utils/` free from core dependencies.

scosman replied:
Ahh, makes sense. Let's do a new /importers folder in adapters? Folks might want to use this in the python API and it will be excluded from docs and harder to import if we don't include it in __init__.py

leonardmq commented on app/web_ui/src/lib/ui/upload_dataset_dialog.svelte:
```diff
@@ -0,0 +1,139 @@
+
+      return true
+    } catch (e) {
+      console.error("Upload failed:", e)
```
✅

leonardmq commented on app/web_ui/src/lib/ui/upload_dataset_dialog.svelte:
```diff
@@ -0,0 +1,139 @@
+
+<Dialog
+  bind:this={dialog}
+  title="Upload Dataset"
```
✅

leonardmq commented on app/web_ui/src/lib/ui/upload_dataset_dialog.svelte:
```diff
@@ -0,0 +1,139 @@
+    {/if}
+  </div>
+  {#if error}
+    <div class="text-error text-sm">
```
✅

leonardmq commented on app/web_ui/src/routes/(app)/dataset/[project_id]/[task_id]/+page.svelte:
```diff
@@ -484,6 +496,11 @@
   sub_subtitle_link="https://docs.getkiln.ai/docs/organizing-datasets"
   no_y_padding
 >
+  <div class="flex flex-row justify-end">
```
✅

leonardmq commented:
> * Let’s put a little helper text in the modal/dialog explaining the fields
>   
>   * “Upload a CSV to add each row to your dataset. The CSV must have a header row. The the following columns are supported: input, output, reasoning (optional), tags (optional, comma separated string).”

I added a helper with the text explaining this and changed to `font-light` to align with similar UIs elsewhere in the code:

<img width="522" alt="image" src="https://github.com/user-attachments/assets/796ed3a8-0366-4a5f-a63f-384e3021d88c" />

scosman commented:
Awesome, will try to 👀  and merge soon!

leonardmq commented:
@scosman - thanks for the review - and there are two comments (marked with 👀 ) that I have not addressed yet but am planning to in the next couple of days:
- adding `chain_of_thought` -> sounds easy to add, but I want to test that out
- adding the validators for the `DataSource` -> I found where this is going on and get the idea, but need to look into the details

leonardmq commented on app/web_ui/src/routes/(app)/dataset/[project_id]/[task_id]/upload_dataset_dialog.svelte:
```diff
@@ -0,0 +1,135 @@
+          </ul>
+          <p class="mb-4">
+            See an example <a
+              href="#link-to-example-on-github"
```
@scosman 

I have a couple of example datasets that demonstrate the structure, I was thinking we can put them somewhere in the repo and link to them from this UI. The CSV structure is easy enough, but might not be obvious how JSON should be presented in it or how to escape the fields.

Is there any directory where we could put those example files?

scosman replied:
I'd say your UI is pretty great. I don't think people will need examples, but if they do the docs are the most discoverable place. I'll add a "Importing CSV" section to [this page](https://docs.getkiln.ai/docs/organizing-datasets), if you want to send some examples I can include them.

leonardmq commented:
All the comments should be addressed now

scosman commented:
Really fantastic work @leonardmq ! Merging.
